manpages/imodextract.man
.na
.nh
.TH imodextract 1 2.20 BL3DEMC
.SH NAME
imodextract \- Extract objects from a model into a new model.
.SH SYNOPSIS
imodextract  list_of_objects  input_model  output_model
.SH DESCRIPTION
Extracts a subset of the objects from an IMOD model and places them into a new 
model.  This can be much more convenient than making a copy of a model and 
deleting objects one by one inside 3dmod.  The list of object numbers
can consist of comma-separated ranges of numbers, for example
1-3,6-9,13,15.
.SH AUTHOR
David Mastronarde,  mast@colorado.edu
.SH SEE ALSO
3dmod(1)
.SH BUGS


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/6/13: Added option to Imodinfo to get contour lengths broken out by
fine-grained color, and made -o option provide object list for all outputs.

12/5/13: Updated the eTomo manpage.  Improved the eTomo help statement.
(Bug# 1764)

12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names

imodutil/imodextract.c
/*
 *  imodextract.c  -  Program to extract a list of objects from a model
 *
 *  Author: David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2006 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 */
/*  $Author$

$Date$

$Revision$

Log at end
*/


#include "imodel.h"
#include "b3dutil.h"

static void usage()
{
  fprintf(stderr, "Usage: imodextract <list of objects> <input model> <output model>\n");
  fprintf(stderr, "       The list of objects can include ranges, e.g. 1-3,6,9,13-15\n");
  exit(3);
}

int main(int argc, char **argv)
{
  Imod *inModel;
  int ob, nob, iview;
  int *list;
  int i, nlist, origsize;
    
  if (argc != 4) usage();
    
  list = parselist(argv[1], &nlist);
  if (!list) {
    fprintf(stderr, "imodextract: Error parsing object list\n");
    exit(1);
  }

  inModel = imodRead(argv[2]);
  if (!inModel) {
    fprintf(stderr, "imodextract: Fatal error reading model\n");
    exit(1);
  }


  /* Make nlist new objects */
  origsize = inModel->objsize;
  for (i = 0; i < nlist; i++)
    imodNewObject(inModel);

  /* Extent the object views then shift all existing objects and their views
     to top */
  imodObjviewComplete(inModel);
  for (ob = origsize - 1; ob >= 0; ob--) {
    imodObjectCopy(&inModel->obj[ob], &inModel->obj[ob+nlist]);
    for (iview = 1; iview < inModel->viewsize; iview++)
      inModel->view[iview].objview[ob + nlist] = 
        inModel->view[iview].objview[ob];
  }

  /* Now copy all of the selected ones into place, copying the view down 
     too */
  for (i = 0; i < nlist; i++) {
    ob = list[i] - 1;
    if (ob < 0 || ob >= origsize) {
      fprintf(stderr, "imodextract: Invalid object number %d\n", ob + 1);
      exit (1);
    }
    imodObjectCopy(&inModel->obj[ob + nlist], &inModel->obj[i]);
    for (iview = 1; iview < inModel->viewsize; iview++)
      inModel->view[iview].objview[i] = 
        inModel->view[iview].objview[ob + nlist];
  }

  /* Delete extra objects by just setting objsize, and the object view sizes,
     set current indexes to -1 to avoid problems */
  inModel->objsize = nlist;
  inModel->cindex.point  = -1;
  inModel->cindex.contour = -1;
  inModel->cindex.object = 0;
  for (iview = 1; iview < inModel->viewsize; iview++)
    inModel->view[iview].objvsize = nlist;
      

  imodBackupFile(argv[argc - 1]);
  if (imodOpenFile(argv[argc - 1], "wb", inModel)) {
    fprintf(stderr, "imodextract: Fatal error opening new model\n");
    exit (1);
  }
  imodWriteFile(inModel);
  exit(0);
}

/*
$Log$
Revision 3.6  2004/11/05 19:05:29  mast
Include local files with quotes, not brackets

Revision 3.5  2004/07/07 19:25:30  mast
Changed exit(-1) to exit(3) for Cygwin

Revision 3.4  2003/10/24 03:05:23  mast
open as binary, strip program name and/or use routine for backup file

Revision 3.3  2003/07/31 21:42:46  mast
Extract object views for the particular objects being extracted

Revision 3.2  2003/02/21 23:18:44  mast
Open output file in binary mode

Revision 3.1  2002/01/28 16:17:27  mast
Fixed report of illegal object number

*/
